src/nn/batch_norm.rs

```diff
@@ -0,0 +1,193 @@
+        }
+    }
+}
+
+
+impl<const C: usize> ResetParams for BatchNorm1D<C> {
+    /// Fills [Self::gamma] with 1s and [Self::beta] with 0s.
+    fn reset_params<R: rand::Rng>(&mut self, _: &mut R) {
+        Cpu::fill(self.gamma.mut_data(), &mut |v| *v = 1.0);
+        Cpu::fill(self.beta.mut_data(), &mut |v| *v = 0.0);
+        Cpu::fill(self.running_mean.mut_data(), &mut |v| *v = 0.0);
+        Cpu::fill(self.running_var.mut_data(), &mut |v| *v = 1.0);
+    }
+}
+
+
+impl<const C: usize> CanUpdateWithGradients for BatchNorm1D<C> {
+    /// Updates [Self::gamma] and [Self::beta].
+    fn update<G: GradientProvider>(&mut self, grads: &mut G, unused: &mut UnusedTensors) {
+        self.gamma.update(grads, unused);
+        self.beta.update(grads, unused);
+    }
+}
+
+
+impl<const C: usize> Module<Tensor1D<C, NoneTape>> for BatchNorm1D<C> {
+    type Output = Tensor1D<C, NoneTape>;
+
+    /// Forward implementation for NoneTape
+    /// Uses the tracked statistics to normalize the input
+    /// Then scale using the gamma and beta parameters
+    fn forward(&self, x: Tensor1D<C, NoneTape>) -> Self::Output {
+        let sd = sqrt(add_scalar(self.running_var.duplicate(), self.epsilon));
+        let x = div(sub(x.duplicate(), &self.running_mean.duplicate()), &sd);
+        add(self.beta.duplicate(), &mul(self.gamma.duplicate(), &x))
+    }
+}
+
+
+impl<const B: usize, const C: usize> Module<Tensor2D<B, C, NoneTape>> for BatchNorm1D<C> {
+    type Output = Tensor2D<B, C, NoneTape>;
+
+    /// Forward implementation for NoneTape
+    /// Uses the tracked statistics to normalize the input
+    /// Then scale using the broadcasted gamma and beta parameters
+    fn forward(&self, x: Tensor2D<B, C, NoneTape>) -> Self::Output {
+        let sd = sqrt(add_scalar(self.running_var.duplicate(), self.epsilon));
+        let x = div(sub(x.duplicate(), &self.running_mean.duplicate().broadcast1()), &sd.broadcast1());
+        add(self.beta.duplicate().broadcast1(), &mul(self.gamma.duplicate().broadcast1(), &x))
+    }
+}
+
+
+impl<const B: usize, const C: usize, const L: usize> Module<Tensor3D<B, C, L, NoneTape>> for BatchNorm1D<C> {
+    type Output = Tensor3D<B, C, L, NoneTape>;
+
+    /// Forward implementation for NoneTape
+    /// Uses the tracked statistics to normalize the input
+    /// Then scale using the broadcasted gamma and beta parameters
+    fn forward(&self, x: Tensor3D<B, C, L, NoneTape>) -> Self::Output {
+        let sd = sqrt(add_scalar(self.running_var.duplicate(), self.epsilon));
+        let x = div(sub(x.duplicate(), &self.running_mean.duplicate().broadcast2()), &sd.broadcast2());
+        add(self.beta.duplicate().broadcast2(), &mul(self.gamma.duplicate().broadcast2(), &x))
+    }
+}
+
+
+impl<const B: usize, const C: usize> Module<Tensor2D<B, C, OwnedTape>> for BatchNorm1D<C> {
+    type Output = Tensor2D<B, C, OwnedTape>;
+
+    // TODO Is this the right thing to do? The input has an OwnedTape
```

Comment:
yeah i think this is a case that isn't possible in pytorch. I think it should stick to the `&self` contract, so it shouldn't change if there's no mut, and then use running statistics.